File: tests/csharp/interfaces_generics/multiple_interfaces.cs
// vybe-test: csharp/interfaces_generics/multiple_interfaces
// origin: languages/csharp/tests/csharp/test_interfaces_generics.rs

using static __Harness;

var d = new Doc { Name = "test" }
;
d.Print();
__P((d.Serialize()).ToString());
__Check("Printing: test\nDOC:test");

interface IPrintable {
    void Print();
}

interface ISerializable {
    string Serialize();
}

class Doc : IPrintable, ISerializable {
    public string Name;
    public void Print() { __P(("Printing: " + Name).ToString()); }
    public string Serialize() { return "DOC:" + Name; }
}

public static class __Harness {
    public static string __buf = "";
    public static void __P(string s) { __buf = __buf + s + "\n"; }
    public static void __Pr(string s) { __buf = __buf + s; }
    public static void __Check(string want) {
        if (__buf != want && __buf != want + "\n") {
            Console.WriteLine("FAIL: want [" + want + "] got [" + __buf + "]");
            throw new Exception("assertion failed");
        }
    }
}

File: tests/php/oop_advanced/multiple_interfaces.php
<?php
// vybe-test: php/oop_advanced/multiple_interfaces
// origin: languages/php/tests/php/test_oop_advanced.rs

function __vybe_check($got, $want) {
    // Match the Rust harness's normalisation: strip \r, then drop trailing
    // newlines (it split on "\n" and popped empty trailing elements).
    $got = str_replace("\r", "", $got);
    $got = rtrim($got, "\n");
    if ($got !== $want) {
        echo "FAIL: want [" . $want . "] got [" . $got . "]\n";
        throw new Exception("assertion failed");
    }
    // Replay the program's own output so running the file by hand still
    // behaves like the program it was extracted from.
    echo $got;
    if ($got !== "") {
        echo "\n";
    }
}

ob_start();

interface Readable {
    public function read(): string;
}
interface Writable {
    public function write(string $data): void;
}
class File implements Readable, Writable {
    private string $content = "";
    public function read(): string { return $this->content; }
    public function write(string $data): void { $this->content .= $data; }
}
$f = new File();
$f->write("hello");
$f->write(" world");
echo $f->read(), "\n";

__vybe_check(ob_get_clean(), "hello world");
